pkg/volume/cinder/cinder.go

```diff
@@ -17,60 +17,72 @@ limitations under the License.
 	"k8s.io/kubernetes/pkg/util/keymutex"
 	"k8s.io/kubernetes/pkg/util/mount"
-	"k8s.io/kubernetes/pkg/util/strings"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
 	"k8s.io/kubernetes/pkg/volume"
-	"k8s.io/kubernetes/pkg/volume/util"
 )
 
-// This is the primary entrypoint for volume plugins.
-func ProbeVolumePlugins() []volume.VolumePlugin {
-	return []volume.VolumePlugin{&cinderPlugin{}}
+const cinderVolumePluginName = "kubernetes.io/cinder"
+
+var _ volume.VolumePlugin = &cinderPlugin{}
+var _ volume.PersistentVolumePlugin = &cinderPlugin{}
+var _ volume.DeletableVolumePlugin = &cinderPlugin{}
+var _ volume.ProvisionableVolumePlugin = &cinderPlugin{}
+var _ volume.AttachableVolumePlugin = &cinderPlugin{}
+
+var _ volume.Mounter = &cinderVolumeMounter{}
+var _ volume.Unmounter = &cinderVolumeUnmounter{}
+var _ volume.Deleter = &cinderVolumeDeleter{}
+var _ volume.Provisioner = &cinderVolumeProvisioner{}
+var _ volume.Attacher = &cinderDiskAttacher{}
+var _ volume.Detacher = &cinderDiskDetacher{}
+
+type cinderVolume struct {
+	volName string
+	podUID  types.UID
+	// Unique identifier of the volume, used to find the disk resource in the provider.
+	pdName string
+	// Filesystem type, optional.
+	fsType string
+	// Specifies the partition to mount
+	// partition string
+	// Specifies whether the disk will be attached as read-only.
+	readOnly bool
+	// Secret that contains the keystone credentials, optional.
+	secretRef *v1.LocalObjectReference
+	// Utility interface that provides API calls to the provider to attach/detach disks.
+	manager cinderManager
+	// Mounter interface that provides system calls to mount the global path to the pod local path.
+	mounter mount.Interface
+	// diskMounter provides the interface that is used to mount the actual block device.
+	blockDeviceMounter mount.Interface
+	plugin             *cinderPlugin
+	volume.MetricsNil
 }
 
-type CinderProvider interface {
-	AttachDisk(instanceID string, diskName string) (string, error)
-	DetachDisk(instanceID string, partialDiskId string) error
-	DeleteVolume(volumeName string) error
-	CreateVolume(name string, size int, vtype, availability string, tags *map[string]string) (volumeName string, err error)
-	GetDevicePath(diskId string) string
-	InstanceID() (string, error)
-	GetAttachmentDiskPath(instanceID string, diskName string) (string, error)
-	DiskIsAttached(diskName, instanceID string) (bool, error)
-	DisksAreAttached(diskNames []string, instanceID string) (map[string]bool, error)
-	ShouldTrustDevicePath() bool
-	Instances() (cloudprovider.Instances, bool)
+func (cd *cinderVolume) GetPath() string {
+	name := cinderVolumePluginName
+	return cd.plugin.host.GetPodVolumeDir(cd.podUID, kstrings.EscapeQualifiedNameForDisk(name), cd.volName)
 }
 
 type cinderPlugin struct {
 	host volume.VolumeHost
 	// Guarding SetUp and TearDown operations
 	volumeLocks keymutex.KeyMutex
+	// For tests purposes, otherwise detected using host:
+	cinderProvider CinderProvider
```

Comment:
`cinderProvider` is specific to the cloud provider implementation. Why stick it in this common `cinderPlugin` struct? Wouldn't it make more sense in `cdManagerCloud`?